pkg/controller/attach_detach.go

```diff
@@ -0,0 +1,665 @@
+	if errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	} else if err != nil {
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	// detachment request
+	if requestedToBeDeleted(azVolumeAttachment) {
+		if azVolumeAttachment.Status.State == v1alpha1.AttachmentPending || azVolumeAttachment.Status.State == v1alpha1.Attached || azVolumeAttachment.Status.State == v1alpha1.AttachmentFailed || azVolumeAttachment.Status.State == v1alpha1.DetachmentFailed {
+			if err := r.triggerDetach(ctx, azVolumeAttachment.Name, azVolumeAttachment); err != nil {
+				return reconcile.Result{Requeue: true}, err
+			}
+		}
+		// attachment request
+	} else if azVolumeAttachment.Status.Detail == nil {
+		if azVolumeAttachment.Status.State == v1alpha1.AttachmentPending || azVolumeAttachment.Status.State == v1alpha1.AttachmentFailed {
+			if err := r.triggerAttach(ctx, azVolumeAttachment.Name, azVolumeAttachment); err != nil {
+				return reconcile.Result{Requeue: true}, err
+			}
+		}
+		// promotion request
+	} else if azVolumeAttachment.Spec.RequestedRole != azVolumeAttachment.Status.Detail.Role {
+		if err := r.promote(ctx, azVolumeAttachment.Name, azVolumeAttachment); err != nil {
+			return reconcile.Result{Requeue: true}, err
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileAttachDetach) triggerAttach(ctx context.Context, attachmentName string, azVolumeAttachment *v1alpha1.AzVolumeAttachment) error {
+	// initialize metadata and update status block
+	conditionFunc := func() (bool, error) {
+		azVolumeAttachment, err := azureutils.GetAzVolumeAttachment(ctx, r.client, r.azVolumeClient, attachmentName, r.namespace, true)
+		if err != nil {
+			klog.Errorf("failed to get AzVolumeAttachment (%s): %v", attachmentName, err)
+			// if failed to get azVolumeAttachment then escape the wait
+			return false, err
+		}
+
+		// deepCopy
+		azVolumeAttachment = azVolumeAttachment.DeepCopy()
+		// Initialize finalizer and add label to the object
+		azVolumeAttachment = r.initializeMeta(ctx, azVolumeAttachment)
+		// Update state to attaching
+		azVolumeAttachment = r.updateState(ctx, azVolumeAttachment, v1alpha1.Attaching)
+
+		// update CRI
+		if err := r.update(ctx, azVolumeAttachment); err != nil {
+			return false, nil
+		}
+		return true, nil
+	}
+	if err := wait.PollImmediate(updateAttemptInterval, updateTimeout, conditionFunc); err != nil {
+		return err
+	}
+
+	// TODO reassess if adding additional finalizer to AzVolume CRI is necessary
+	// // add finalizer to the bound AzVolume CRI
+	// conditionFunc = func() (bool, error) {
+	// 	if err := r.addFinalizerToAzVolume(ctx, azVolumeAttachment.Spec.UnderlyingVolume); err != nil {
+	// 		return false, nil
+	// 	}
+	// 	return true, nil
+	// }
+	if err := wait.PollImmediate(updateAttemptInterval, updateTimeout, conditionFunc); err != nil {
+		return err
+	}
+
+	// initiate goroutine to attach volume
+	go func() {
```

Comment:
IIRC, the reconcile function is already called in a goroutine, and there is some concurrency allowed by the controller-runtime. Do we really need to spawn another goroutine to handle the attach especially since operations a not serialized or bounded per attachment?

Reply:
There is a limit to number of maximum concurrent reconciles (we have set it to be 10 for our controllers for now I believe). If we don't call attach or detach in a separate goroutine, this will block other reconcile to kick in, thereby capping maximum concurrent attach / detach operations to whatever value we set for `MaxmiumConcurrentReconciles`. And this might hurt our performance if a large amount of attach / detach operations kick in.

Reply:
Fair enough. Still, I wonder if we should at least serial attach/detach operations per `AzVolumeAttachment` instance to limit concurrency since attach & detach operations can take a long time. If the CSI timeout is set too short, the operation may be retried while the previous is still pending resulting in a failure.

Reply:
That makes sense. The current reconciliation loop relies on `AzVolumeAttachmentState` to handle that. It might run into abovementioned occurrences if a reconciliation for the same `AzVolumeAttachment` instance kicks in before `AzVolumeAttachmentState` is updated but I wonder how often that will happen given that crdProvisioner will only update existing `AzVolumeAttachment` CRI if `AzVolumeAttachment.Status.Error` is observed.

https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/669992ab3721f782b05a71b73e3c22fd4113e252/pkg/controller/attach_detach.go#L86
https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/669992ab3721f782b05a71b73e3c22fd4113e252/pkg/controller/attach_detach.go#L79

Reply:
Ah, I see. In theory, that scenario may be possible. In practice, not very likely. This should be sufficient.